### PR TITLE
feature: Finish exec plugin support by wiring up support in plugins and runners

### PR DESCRIPTION
**Pull request by evanphx**
This depends on https://github.com/hashicorp/waypoint-plugin-sdk/pull/10 as it adds the required plugin support to be able to have plugin components that can exec.

This wires the plugin support end-to-end by adding support for exec plugins as runner jobs and for the exec cli command to spawn those jobs and connect back to them.

### CODE REVIEW

File: internal/client/job.go [hunk 1]
```diff
@@ -50,9 +50,18 @@ func (c *Project) job() *pb.Job {
 
 // doJob will queue and execute the job. If the client is configured for
 // local mode, this will start and target the proper runner.
-func (c *Project) doJob(ctx context.Context, job *pb.Job, ui terminal.UI) (*pb.Job_Result, error) {
+// the mon channel can be used to monitor the job status as it changes.
+// The receiver must be careful to not block sending to mon as it will block
+// the job state processing loop.
+func (c *Project) doJob(ctx context.Context, job *pb.Job, ui terminal.UI, mon chan pb.Job_State) (*pb.Job_Result, error) {
```
Comment (mitchellh) [inline internal/client/job.go:56]: Given how many things call `doJob` and don't care about `mon`, I feel like it'd be neater to do a second `doJobMonitor` function that has this signature and have `doJob` call into that with the preserved previous signature. This is a common pattern in Go libs.

File: internal/client/job.go [hunk 2]
```diff
@@ -50,9 +50,18 @@ func (c *Project) job() *pb.Job {
...
 	log := c.logger
 
+	// Be sure that the monitor is closed so the reciever knows for sure the job isn't going
+	// anymore.
+	if mon != nil {
```
Comment (mitchellh) [inline internal/client/job.go:61]: Also, nitpick but I like it when channels end in `Ch` so I know they're channels. Its a LITTLE weird but seems like fairly idiomatic too?

File: internal/client/job.go [hunk 3]
```diff
@@ -114,6 +123,7 @@ func (c *Project) queueAndStreamJob(
 	job *pb.Job,
 	ui terminal.UI,
 	jobIdCallback func(string),
+	mon chan pb.Job_State,
```
Comment (mitchellh) [inline internal/client/job.go:126]: This is fine, cause nothing calls `queueAndStreamJob` directly except `doJob` if I recall.

File: internal/core/app_exec.go [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/hashicorp/waypoint-plugin-sdk/component"
+	"github.com/hashicorp/waypoint/internal/ceb"
+	"github.com/hashicorp/waypoint/internal/config"
+	pb "github.com/hashicorp/waypoint/internal/server/gen"
+)
+
+// Exec launches an exec plugin
```
Comment (mitchellh) [inline internal/core/app_exec.go:21]: I'd try to add a lot more detail here as it relates to CEB stuff, just so its clear this is only for _plugin based exec_ and not general CEB-enabled exec.

File: internal/core/app_exec.go [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+func (a *App) Exec(ctx context.Context, id string, d *pb.Deployment) error {
+	// We need to get the pushed artifact if it isn't loaded.
+	// TODO(evanphx) I don't understand why we do this. We're mimicing what the
+	// deploy destroy code does and they load this, so we do to?
```
Comment (mitchellh) [inline internal/core/app_exec.go:25]: To address this comment: this is for [this feature](https://www.waypointproject.io/docs/waypoint-hcl/variables/artifact) (and similar).

File: internal/core/app_exec.go [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+	_, _, err = a.doOperation(ctx, a.logger.Named("exec"), &execOperation{
+		InstanceId: id,
+		Component:  c,
+		Deployment: d,
+		Client:     a.client,
+	})
+	return err
+}
+
+// execOperation is a simple barebones operation value to conform to the
```
Comment (mitchellh) [inline internal/core/app_exec.go:80]: Do you need an operation? Or can you just call it directly?

File: internal/core/app_exec.go [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+	p.log.Debug("calling plugin with session-info", "arguments", esi.Arguments)
+
+	val, err := p.app.callDynamicFunc(ctx,
+		p.log,
+		nil,
+		p.op.Component,
+		p.execer.ExecFunc(),
+		argNamedAny("deployment", p.op.Deployment.Deployment),
+		argmapper.Named("exec_info", esi),
```
Comment (mitchellh) [inline internal/core/app_exec.go:181]: Can also do a `argmapper.Typed` here. It doesn't make a big difference but if you don't plan to ever have multiple `ExecSessionInfo` args then its probably less confusing.

File: internal/server/execclient/client.go [hunk 8]
```diff
@@ -142,7 +142,7 @@ func (c *Client) Run() (int, error) {
 
 	if ptyF != nil {
 		// We need to go into raw mode with stdin
-		if f, ok := c.Stdin.(*os.File); ok {
+		if f, ok := c.Stdout.(*os.File); ok {
```
Comment (mitchellh) [inline internal/server/execclient/client.go:145]: This makes sense but also seems suspect. We have been using this code for some time with ptys with no problem.(even executing stuff like `vim` and resizing windows). Why this change?

File: internal/server/proto/server.proto [hunk 9]
```diff
@@ -783,6 +787,17 @@ message Job {
     Deployment deployment = 1;
   }
 
+  message ExecOp {
```
Comment (mitchellh) [inline internal/server/proto/server.proto:790]: Would comment again that this is for plugin-powered exec and not CEB

File: internal/server/proto/server.proto [hunk 10]
```diff
@@ -58,6 +58,9 @@ service Waypoint {
   // ListInstances returns the running instances of deployments.
   rpc ListInstances(ListInstancesRequest) returns (ListInstancesResponse);
 
+  // FindExecInstance returns an instance that can be used for an exec session.
```
Comment (mitchellh) [inline internal/server/proto/server.proto:61]: Actually not sure what the use case is for this, can you address that in the comment?

File: internal/server/singleprocess/state/instance_exec.go [hunk 11]
```diff
@@ -97,6 +98,71 @@ func (s *State) InstanceExecCreateByTargetedInstance(id string, exec *InstanceEx
 	return nil
 }
 
+// CalculateInstanceExecByDeployment considers all the instances registered
+// to the given deployment and finds the one that is least loaded. If there
+// are no instances, returns a ResourceExhausted error. Calls to this
+// function in quick succession will return could return the same instance,
+// which is why a simple random sampling is done on all prospective instances.
+func (s *State) CalculateInstanceExecByDeployment(did string) (*Instance, error) {
+	txn := s.inmem.Txn(true)
```
Comment (mitchellh) [inline internal/server/singleprocess/state/instance_exec.go:107]: I think this can be `false` cause its a read-only transaction, yeah?

File: internal/server/singleprocess/state/instance_exec.go [hunk 12]
```diff
@@ -97,6 +98,71 @@ func (s *State) InstanceExecCreateByTargetedInstance(id string, exec *InstanceEx
 	return nil
 }
 
+// CalculateInstanceExecByDeployment considers all the instances registered
+// to the given deployment and finds the one that is least loaded. If there
+// are no instances, returns a ResourceExhausted error. Calls to this
+// function in quick succession will return could return the same instance,
+// which is why a simple random sampling is done on all prospective instances.
+func (s *State) CalculateInstanceExecByDeployment(did string) (*Instance, error) {
```
Comment (mitchellh) [inline internal/server/singleprocess/state/instance_exec.go:106]: This should have unit tests and can surely be well unit tested.

**Review by mitchellh — CHANGES_REQUESTED** [review]

File: internal/server/execclient/client.go (hunk 8 above)
Comment (evanphx) [reply internal/server/execclient/client.go:145]: The commit message on that line explains the bug. Basically it has to do with someone doing `date | waypoint exec`.

File: internal/core/app_exec.go (hunk 5 above)
Comment (evanphx) [reply internal/core/app_exec.go:25]: Gotcha, thanks!

File: internal/core/app_exec.go (hunk 6 above)
Comment (evanphx) [reply internal/core/app_exec.go:80]: So I was wondering that and I'd love to chat about that with you. I was mimic'ing the other functions and they all seemed to use an operation.

File: internal/core/app_exec.go (hunk 7 above)
Comment (evanphx) [reply internal/core/app_exec.go:181]: Ah, gotcha.  👍🏻

File: internal/server/execclient/client.go (hunk 8 above)
Comment (mitchellh) [reply internal/server/execclient/client.go:145]: Hm hm yes. Okay, just verify `vim` and friends still work.

File: internal/client/noop.go [hunk 13]
```diff
@@ -21,6 +21,6 @@ func (c *App) Noop(ctx context.Context) error {
 	}
 
 	// Execute it
-	_, err := c.doJob(ctx, job)
+	_, err := c.doJob(ctx, job, nil)
```
Comment (mitchellh) [inline internal/client/noop.go:24]: This is probably wrong, not compiling?

**Review by mitchellh — CHANGES_REQUESTED** [review]
Just gotta fix the failing tests but this looks good.

File: internal/client/noop.go (hunk 13 above)
Comment (evanphx) [reply internal/client/noop.go:24]: Oh strange! I fixed those but forgot to push the files. Should be in now.

**Review by mitchellh — APPROVED** [review]
